File: software/drivers/motors-mirroring/src/inference/mirror.rs
#![allow(clippy::collapsible_if)]

use crate::{config::MotorConfig, inference::model::{MotorState, MotorProtectionState, MovementDirection, ProtectionKey}, r#types::{Command, MotorCommand}};
use super::normalize;
use std::collections::HashMap;

fn calculate_shortest_distance(pos1: u16, pos2: u16, config: &MotorConfig) -> u16 {
    let half_max_steps = config.max_steps / 2;
    let distance = pos1.abs_diff(pos2);

    if distance > half_max_steps {
        config.max_steps - distance
    } else {
        distance
    }
}

fn get_speed_and_accel_by_distance(
    target_bus_id: &str,
    motor_id: u16,
    distance: u16,
    state: &MotorState,
    config: &MotorConfig,
    commands: &mut Vec<Command>,
) {
    let share = distance as f64 / (config.max_steps / 2) as f64;
    let speed =
        (share * (config.max_speed - config.min_speed) as f64) as u16 + config.min_speed;
    let accel =
        (share * (config.max_accel - config.min_accel) as f64) as u16 + config.min_accel;

    if speed == state.goal_speed && (accel as u8) == state.goal_accel {
        return;
    }

    commands.extend_from_slice(&[
        Command {
            target_bus_id: target_bus_id.to_string(),
            motor_id: motor_id as u32,
            command: MotorCommand::Speed(speed),
        },
        Command {
            target_bus_id: target_bus_id.to_string(),
            motor_id: motor_id as u32,
            command: MotorCommand::Accel(accel),
        },
    ]);
}

pub fn get_movement_sequence_with_goal(
    target_bus_id: &str,
    motor_id: u16,
    source: &MotorState,
    target: &MotorState,
    config: &MotorConfig,
    protection_states: &mut HashMap<ProtectionKey, MotorProtectionState>,
    commands: &mut Vec<Command>,
) {
    let source_percent = normalize::normalize_position(
        source.present_position,
        source.range_min,
        source.range_max,
        config,
    );

    // Voltage faults on the leader are common (rail dips under load) and shouldn't
    // stop mirroring — mask bit 0 before gating.
    const MIRRORING_IGNORED_BITS: u8 = 0x01; // Voltage
    let blocking_source = source.error_status & !MIRRORING_IGNORED_BITS;
    if blocking_source != 0 {
        let source_errors = st3215::protocol::ServoError::from_bits(source.error_status);
        let target_errors = st3215::protocol::ServoError::from_bits(target.error_status);
        log::warn!(
            "Skipping mirroring for motor {} on bus {} due to error state [source=0x{:02X} {:?}, target=0x{:02X} {:?}], curr = {}, limit = {}",
            motor_id, target_bus_id,
            source.error_status, source_errors,
            target.error_status, target_errors,
            source.current, source.current_limit
        );
        return;
    }
    if source.present_position == 0 || target.present_position == 0 {
        // just skip for now
        log::warn!("Skipping mirroring for motor {} on bus {} due to zero position [source={}, target={}]", motor_id, target_bus_id, source.present_position, target.present_position);
        return;
    }

    let target_range_size =
        normalize::get_steps_range(target.range_min, target.range_max, config);
    if target_range_size < config.safety_margin * 2 {
        log::warn!("Skipping mirroring for motor {} on bus {} due to insufficient range size", motor_id, target_bus_id);
        return;
    }
    let target_range_size = target_range_size - config.safety_margin * 2;

    let goal_offset = (source_percent / 100.0 * target_range_size as f64) as u16 + config.safety_margin;
    let goal_position = ((target.range_min as u32 + goal_offset as u32)
        & (config.max_steps - 1) as u32) as u16;

    // Get or create protection state for this motor
    let protection_key = (target_bus_id.to_string(), motor_id as u8);
    let protection_state = protection_states.entry(protection_key).or_default();

    // Determine what direction the leader is commanding
    let leader_direction = if goal_position > target.present_position {
        Some(MovementDirection::Positive)
    } else if goal_position < target.present_position {
        Some(MovementDirection::Negative)
    } else {
        None // No movement needed
    };

    // Current protection: track direction that causes high current and block it
    let current_threshold = config.get_current_threshold(motor_id as u8);
    if current_threshold > 0 {
        if target.current >= current_threshold {
            // Current is high - determine which direction caused it if not already known
            if protection_state.blocked_direction.is_none() && protection_state.previous_position != target.present_position {
                let movement_direction = if target.present_position > protection_state.previous_position {
                    MovementDirection::Positive
                } else {
                    MovementDirection::Negative
                };

                protection_state.blocked_direction = Some(movement_direction);
                log::warn!(
                    "Motor {} on bus {} current {} exceeds threshold {}, blocking {:?} direction",
                    motor_id, target_bus_id, target.current, current_threshold, movement_direction
                );
            }

            // Hold current position when current is high
            if target.target_position.abs_diff(target.present_position) > config.deadband {
                let goal_cmd = Command {
                    target_bus_id: target_bus_id.to_string(),
                    motor_id: motor_id as u32,
                    command: MotorCommand::Goal(target.present_position),
                };
                commands.push(goal_cmd);
            }
            return;
        } else {
            // Current is normal - only clear blocked direction if leader is commanding opposite movement
            if let Some(blocked_dir) = protection_state.blocked_direction {
                if let Some(leader_dir) = leader_direction {
                    if leader_dir != blocked_dir {
                        // Leader wants to move in opposite direction - safe to clear protection
                        log::info!(
                            "Motor {} on bus {} current {} back to normal and leader commanding opposite direction, clearing protection",
                            motor_id, target_bus_id, target.current
                        );
                        protection_state.blocked_direction = None;
                    } else {
                        // Leader still wants to move in blocked direction - keep protection active
                        log::warn!(
                            "Motor {} on bus {} current normal but leader still commanding blocked {:?} direction, maintaining protection",
                            motor_id, target_bus_id, blocked_dir
                        );
                    }
                }
            }
        }
    }

    // Check if intended movement is in blocked direction
    if let Some(blocked_dir) = protection_state.blocked_direction {
        if let Some(intended_direction) = leader_direction {
            if intended_direction == blocked_dir {
                log::warn!(
                    "Motor {} on bus {} blocking {:?} movement (would increase current)",
                    motor_id, target_bus_id, blocked_dir
                );
                protection_state.previous_position = target.present_position;
                return;
            } else {
                log::info!(
                    "Motor {} on bus {} allowing {:?} movement (opposite to blocked {:?}, may relieve current)",
                    motor_id, target_bus_id, intended_direction, blocked_dir
                );
            }
        } else {
            // No movement needed
            protection_state.previous_position = target.present_position;
            return;
        }
    }

    let distance = calculate_shortest_distance(target.present_position, goal_position, config);

    if distance <= config.deadband {
        return;
    }
    if target.target_position.abs_diff(goal_position) <= config.deadband {
        return;
    }

    get_speed_and_accel_by_distance(target_bus_id, motor_id, distance, target, config, commands);

    let goal_cmd = Command {
        target_bus_id: target_bus_id.to_string(),
        motor_id: motor_id as u32,
        command: MotorCommand::Goal(goal_position),
    };

    commands.push(goal_cmd);

    // Update previous position for next iteration's direction tracking
    let protection_key = (target_bus_id.to_string(), motor_id as u8);
    if let Some(protection_state) = protection_states.get_mut(&protection_key) {
        protection_state.previous_position = target.present_position;
    }
}

File: software/station/examples/st3215-remote-teleop-py/mirror.py
"""Per-motor goal computation: leader position -> follower command.

For each motor pair (same id on leader and follower bus) we:
  1. Convert the leader's present position to a 0-100% within its calibrated arc.
  2. Project that percentage onto the follower's calibrated arc (with safety margin).
  3. Apply current-overload protection: if the follower drew too much current
     while moving, block further movement in that direction until the leader
     reverses course.
  4. Pick speed/accel proportional to the remaining travel distance.
"""

from dataclasses import dataclass, field
from enum import Enum

from config import MotorConfig


# Motor status register bit 0 = under-voltage. Common on leaders when their rail
# dips under load and not a real fault — mask it out before gating.
STATUS_VOLTAGE_BIT = 0x01


class Direction(Enum):
    POSITIVE = 1
    NEGATIVE = -1


@dataclass
class MotorState:
    """Snapshot of one motor extracted from a single inference frame."""
    present_position: int = 0
    target_position: int = 0
    range_min: int = 0
    range_max: int = 0
    error_status: int = 0
    current: int = 0
    goal_speed: int = 0
    goal_accel: int = 0


@dataclass
class ProtectionState:
    """Per-motor memory across teleop ticks: which direction is currently blocked."""
    blocked_direction: Direction | None = None
    previous_position: int = 0


@dataclass
class MotorCommand:
    """One follower-bus write the teleop loop wants to issue this tick."""
    motor_id: int
    speed: int | None = None
    accel: int | None = None
    goal: int | None = None


def get_steps_range(range_min: int, range_max: int, config: MotorConfig) -> int:
    """Width of the calibrated arc in encoder steps, accounting for wrap-around."""
    if range_max >= range_min:
        return range_max - range_min
    return (config.max_steps - range_min) + range_max


def normalize_position(
    position: int, range_min: int, range_max: int, config: MotorConfig,
) -> float:
    """Map a raw encoder position to a 0-100% within [range_min, range_max].

    Handles wrap-around arcs (where range_max < range_min, e.g. the arc spans
    the encoder's zero crossing). Out-of-range positions clamp to the nearest
    end. Returns 50% when the arc has zero width (uncalibrated).
    """
    range_size = get_steps_range(range_min, range_max, config)
    if range_size == 0:
        return 50.0

    if range_max >= range_min:
        if position < range_min:
            relative = 0
        elif position > range_max:
            relative = range_size
        else:
            relative = position - range_min
    else:
        # Wrap-around arc: valid region is [range_min, max_steps) U [0, range_max].
        if position >= range_min:
            relative = position - range_min
        elif position <= range_max:
            relative = (config.max_steps - range_min) + position
        else:
            # Outside the arc — clamp to whichever end is closer.
            dist_to_min = range_min - position
            dist_to_max = position - range_max
            relative = 0 if dist_to_min < dist_to_max else range_size

    pct = (relative / range_size) * 100.0
    return max(0.0, min(100.0, pct))


def shortest_distance(a: int, b: int, config: MotorConfig) -> int:
    """Shortest distance between two encoder positions, going either direction."""
    half = config.max_steps // 2
    d = abs(a - b)
    return config.max_steps - d if d > half else d


def speed_and_accel_for_distance(
    distance: int, config: MotorConfig,
) -> tuple[int, int]:
    """Pick speed/accel proportional to how far the follower has to travel."""
    share = distance / (config.max_steps / 2)
    speed = int(share * (config.max_speed - config.min_speed)) + config.min_speed
    accel = int(share * (config.max_accel - config.min_accel)) + config.min_accel
    return speed, accel


def compute_motor_command(
    motor_id: int,
    leader: MotorState,
    follower: MotorState,
    protection: ProtectionState,
    config: MotorConfig,
) -> MotorCommand | None:
    """Decide what to send to one follower motor for this tick.

    Returns None when no command is needed (deadband, stale data, etc.).
    Mutates `protection` to remember the direction tracking across ticks.
    """
    # Skip if leader has a real fault (voltage dips don't count).
    if (leader.error_status & ~STATUS_VOLTAGE_BIT) != 0:
        return None

    # Zero positions usually mean the motor hasn't reported yet.
    if leader.present_position == 0 or follower.present_position == 0:
        return None

    # Need enough room on the follower's arc to apply the safety margin.
    follower_range = get_steps_range(follower.range_min, follower.range_max, config)
    if follower_range < config.safety_margin * 2:
        return None
    usable_range = follower_range - config.safety_margin * 2

    # Map leader's position to follower's coordinate system.
    leader_pct = normalize_position(
        leader.present_position, leader.range_min, leader.range_max, config,
    )
    goal_offset = int(leader_pct / 100.0 * usable_range) + config.safety_margin
    goal_position = (follower.range_min + goal_offset) % config.max_steps

    # What direction is the leader asking for?
    if goal_position > follower.present_position:
        leader_direction: Direction | None = Direction.POSITIVE
    elif goal_position < follower.present_position:
        leader_direction = Direction.NEGATIVE
    else:
        leader_direction = None

    threshold = config.get_current_threshold(motor_id)
    if threshold > 0:
        if follower.current >= threshold:
            # Current is high. Figure out which direction caused it (by comparing
            # to the previous tick's position) and block that direction.
            if (
                protection.blocked_direction is None
                and protection.previous_position != follower.present_position
            ):
                if follower.present_position > protection.previous_position:
                    protection.blocked_direction = Direction.POSITIVE
                else:
                    protection.blocked_direction = Direction.NEGATIVE

            # Hold position while current is elevated. Only re-issue the goal
            # if the follower is meaningfully off its current target.
            if abs(follower.target_position - follower.present_position) > config.deadband:
                return MotorCommand(motor_id=motor_id, goal=follower.present_position)
            return None

        # Current is back to normal. Clear the block only if the leader is now
        # asking for movement in the opposite direction (the original cause is
        # likely still mechanically present).
        if (
            protection.blocked_direction is not None
            and leader_direction is not None
            and leader_direction != protection.blocked_direction
        ):
            protection.blocked_direction = None

    # Leader still wants to push into the blocked direction — refuse.
    if protection.blocked_direction is not None:
        if leader_direction is None or leader_direction == protection.blocked_direction:
            protection.previous_position = follower.present_position
            return None

    distance = shortest_distance(follower.present_position, goal_position, config)

    # Already close enough — don't send anything.
    if distance <= config.deadband:
        return None
    if abs(follower.target_position - goal_position) <= config.deadband:
        return None

    cmd = MotorCommand(motor_id=motor_id, goal=goal_position)

    # Only emit speed/accel when they would actually change.
    speed, accel = speed_and_accel_for_distance(distance, config)
    if speed != follower.goal_speed:
        cmd.speed = speed
    if accel != follower.goal_accel:
        cmd.accel = accel

    protection.previous_position = follower.present_position
    return cmd


@dataclass
class ProtectionTable:
    """Persistent protection state for every follower motor we've seen."""
    states: dict[int, ProtectionState] = field(default_factory=dict)

    def get(self, motor_id: int) -> ProtectionState:
        if motor_id not in self.states:
            self.states[motor_id] = ProtectionState()
        return self.states[motor_id]
